pointcloud/octree/basic.go

```diff
@@ -0,0 +1,171 @@
+	tree     []*basicOctree
+	point    pc.PointAndData
+}
+
+// New creates a new basic octree with specified center, side and metadata.
+func New(ctx context.Context, center r3.Vector, sideLength float64, logger golog.Logger) (Octree, error) {
+	if sideLength <= 0 {
+		return nil, errors.Errorf("invalid side length (%.2f) for octree", sideLength)
+	}
+
+	octree := &basicOctree{
+		ctx:    ctx,
+		logger: logger,
+		node:   newLeafNodeEmpty(),
+		center: center,
+		side:   sideLength,
+	}
+	octree.meta = octree.NewMetaData()
+	return octree, nil
+}
+
+// Size returns the number of points stored in the octree by traversing it and incrementing based on the number
+// of nodes containing a point.
+func (octree *basicOctree) Size() int {
+	var totalSize int
+
+	switch octree.node.nodeType {
+	case InternalNode:
+		for _, children := range octree.node.tree {
+			totalSize += children.Size()
+		}
+
+	case LeafNodeFilled:
+		totalSize = 1
+
+	case LeafNodeEmpty:
+		totalSize = 0
+	}
+
+	return totalSize
+}
+
+// Set checks if the point to be added is a valid point for the OCtree to hold based on its center and side length.
+// It then recursively iterates through the tree until it finds the appropriate node to add it too. If the found node
+// contains a point already, it will the node into octants and add both the old and new points to them respectively.
+func (octree *basicOctree) Set(p r3.Vector, d pc.Data) error {
+	if (pc.PointAndData{P: p, D: d} == pc.PointAndData{}) {
+		octree.logger.Debug("no data given, skipping insertion")
+		return nil
+	}
+
+	if !checkPointPlacement(octree.center, octree.side, p) {
+		return errors.New("error point is outside the bounds of this octree")
+	}
+
+	switch octree.node.nodeType {
+	case InternalNode:
+		for _, childNode := range octree.node.tree {
+			if checkPointPlacement(childNode.center, childNode.side, p) {
+				// Iterate through children nodes
+				err := childNode.Set(p, d)
+				if err == nil {
+					// Update metadata
+					octree.meta.Merge(p, d)
+				}
+				return err
+			}
+		}
+		return errors.New("error invalid internal node detected, please check your tree")
+
+	case LeafNodeFilled:
+		_, exists := octree.At(p.X, p.Y, p.Z)
+		if exists {
+			// Update data in point
+			octree.node.point.D = d
+			return nil
+		}
+		err := octree.splitIntoOctants()
+		if err != nil {
+			return errors.Errorf("error in splitting octree into new octants: %v", err)
+		}
+		// No update of metadata as the set call below will lead to the InternalNode case due to the octant split
+		return octree.Set(p, d)
+
+	case LeafNodeEmpty:
+		// Update metadata
+		octree.meta.Merge(p, d)
+		octree.node = newLeafNodeFilled(p, d)
+	}
+
+	return nil
+}
+
+// At traverses the octree to see in a point exists at the specified location. If a point does exist, its data is
+// returned along with true, if one is not then no data is returned and the boolean is returned false.
+func (octree *basicOctree) At(x, y, z float64) (pc.Data, bool) {
```

Comment:
As discussed, current implementation is for exact match to return true and the point data. nearest neighbor has not been implemented